File: transpiler/src/main/java/com/here/ivi/api/generator/cpp/CppGenerator.java
/*
 * Copyright (C) 2017 HERE Global B.V. and its affiliate(s). All rights reserved.
 *
 * This software, including documentation, is protected by copyright controlled by
 * HERE Global B.V. All rights are reserved. Copying, including reproducing, storing,
 * adapting or translating, any or all of this material requires the prior written
 * consent of HERE Global B.V. This material also contains confidential information,
 * which may not be disclosed to others without prior written consent of HERE Global B.V.
 *
 */

package com.here.ivi.api.generator.cpp;

import com.here.ivi.api.generator.common.GeneratedFile;
import com.here.ivi.api.generator.common.TemplateEngine;
import com.here.ivi.api.model.cppmodel.CppFile;
import java.io.File;
import java.util.*;

public final class CppGenerator {
  public List<GeneratedFile> generateCode(
      final CppFile cppModel,
      final String relativeHeaderPath,
      final String relativeImplPath,
      final String pathPrefix) {

    if (cppModel == null || cppModel.isEmpty()) {
      return Collections.emptyList();
    }

    String absoluteHeaderPath =
        pathPrefix
            + File.separator
            + CppNameRules.PACKAGE_NAME_SPECIFIER_INCLUDE
            + File.separator
            + relativeHeaderPath
            + CppNameRules.HEADER_FILE_SUFFIX;
    String absoluteImplPath =
        pathPrefix
            + File.separator
            + CppNameRules.PACKAGE_NAME_SPECIFIER_SRC
            + File.separator
            + relativeImplPath
            + CppNameRules.IMPLEMENTATION_FILE_SUFFIX;

    // Filter out self-includes
    cppModel.includes.removeIf(
        include -> include.fileName.equals(relativeHeaderPath + CppNameRules.HEADER_FILE_SUFFIX));

    String commentHeader = TemplateEngine.render("cpp/CppCommentHeader", null);

    List<GeneratedFile> result = new LinkedList<>();
    String headerContent = TemplateEngine.render("cpp/CppHeader", cppModel);
    result.add(new GeneratedFile(commentHeader + headerContent, absoluteHeaderPath));

    String headerInclude =
        "\n#include \"" + (relativeHeaderPath + CppNameRules.HEADER_FILE_SUFFIX) + "\"";
    String implementationContent = TemplateEngine.render("cpp/CppImplementation", cppModel);
    result.add(
        new GeneratedFile(commentHeader + headerInclude + implementationContent, absoluteImplPath));

    return result;
  }
}

File: genium/src/main/java/com/here/genium/generator/cpp/CppGenerator.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cpp

import com.here.genium.generator.common.GeneratedFile
import com.here.genium.generator.common.templates.TemplateEngine
import com.here.genium.model.common.Include
import com.here.genium.model.cpp.CppConstant
import com.here.genium.model.cpp.CppExternableElement
import com.here.genium.model.cpp.CppFile
import com.here.genium.model.cpp.CppUsing
import java.nio.file.Paths

class CppGenerator(private val pathPrefix: String, private val internalNamespace: List<String>) {

    fun generateCode(cppModel: CppFile): List<GeneratedFile> {

        val hasConstants = cppModel.members.any { it is CppConstant }
        val hasTypedefs = cppModel.members.any { it is CppUsing }
        val hasNonExternalElements =
            cppModel.members.filterIsInstance<CppExternableElement>().any { !it.isExternal }
        val hasErrorEnums = cppModel.errorEnums.isNotEmpty()
        val hasCode = hasConstants || hasNonExternalElements || hasErrorEnums || hasTypedefs
        if (!hasCode) {
            return emptyList()
        }

        val relativeFilePath = cppModel.filename
        val absoluteHeaderPath = Paths.get(
            pathPrefix,
            PACKAGE_NAME_SPECIFIER_INCLUDE,
            relativeFilePath
        ).toString() + HEADER_FILE_SUFFIX
        val absoluteImplPath = Paths.get(
            pathPrefix,
            PACKAGE_NAME_SPECIFIER_SRC,
            relativeFilePath
        ).toString() + IMPLEMENTATION_FILE_SUFFIX

        // Filter out self-includes
        val includes = cppModel.includes
        includes.removeIf { it.fileName == relativeFilePath + HEADER_FILE_SUFFIX }
        CppLibraryIncludes.filterIncludes(includes, internalNamespace)

        val result = mutableListOf<GeneratedFile>()

        val headerContent = TemplateEngine.render("cpp/CppHeader", cppModel)
        result.add(GeneratedFile(headerContent, absoluteHeaderPath))

        cppModel.headerInclude =
            Include.createInternalInclude(relativeFilePath + HEADER_FILE_SUFFIX)

        val implementationContent = TemplateEngine.render("cpp/CppImplementation", cppModel)
        result.add(GeneratedFile(implementationContent, absoluteImplPath))

        return result
    }

    fun generateHelperHeader(headerName: String) =
        generateHelperHeader(headerName, mapOf("internalNamespace" to internalNamespace))

    fun generateHelperHeader(headerName: String, extraInfo: Any): GeneratedFile {
        val content = TemplateEngine.render("cpp/$headerName", extraInfo)
        val resultFileName = Paths.get(
            pathPrefix,
            PACKAGE_NAME_SPECIFIER_INCLUDE,
            internalNamespace.joinToString("/"),
            headerName
        ).toString() + HEADER_FILE_SUFFIX
        return GeneratedFile(content, resultFileName)
    }

    companion object {
        const val HEADER_FILE_SUFFIX = ".h"
        private const val IMPLEMENTATION_FILE_SUFFIX = ".cpp"
        private const val PACKAGE_NAME_SPECIFIER_INCLUDE = "include"
        private const val PACKAGE_NAME_SPECIFIER_SRC = "src"
    }
}
